Upload logs when dependencies fail to start

software-opal opened the pull request:
This PR fixes #327 by uploading docker-compose logs when the services fail to start up.

I've done this by copy-pasting the log upload code from the end of the `run.sh` file to below the code that starts up the dependent services.

I've verified that these changes correctly fix the issue using the BuildKite pipeline and docker-compose file in the issue.

toote commented:
Great idea! I would suggest 2 changes, though:

1- instead of copying and pasting the code from the `run.sh` file, move it to the `shared.sh` file inside of a function and call it from both places (thus avoiding code duplication)
2- instead of messing with the `-e` option I believe you shoud just enclose the `run_docker_compose` calls in an `if ! run_docker_compose ...`. I believe that the end result will be the same and the code will be cleaner (avoiding the need to explain why you are doing something that looks wrong :P)

software-opal commented:
Thanks for the feedback, I've gone back and made some changes. I've moved the failed services table and log upload into functions in the `run.bash` shared file. I've also added a test for the 'dependencies fail to start' scenario.

I made some other changes to the tests in `tests/output.bats` because I wasn't confident that my changes were working correctly and I realised the output-checking parts of the test could have passed without the plugin actually running the test cases(the substring being checked for was present in the dependency output too).

toote commented on tests/output.bats:
```diff
@@ -113,8 +158,8 @@ load '../lib/run'
+    "-f docker-compose.yml -p buildkite1111 run --name buildkite1111_myservice_build_1 --rm myservice echo 'hello world' : echo ran myservice command"
 
   stub docker \
     "ps -a --filter label=com.docker.compose.project=buildkite1111 -q : echo" \
```
in the next line there is also an issue with the stub:
```
bats-mock(docker): got docker ps -a --filter label=com.docker.compose.project=buildkite1111 --format {{.ID}}\t{{.Label "com.docker.compose.service"}}
bats-mock(docker): patterns  [3] = 'ps' '-a' '--filter' 
bats-mock(docker): arguments [6] = 'ps' '-a' '--filter' 'label=com.docker.compose.project=buildkite1111' '--format' '\{\{.ID\}\}\\t\{\{.Label\ \"com.docker.compose.service\"\}\}' 
bats-mock(docker): unexpected argument 'label=com.docker.compose.project=buildkite1111' at idx 3
bats-mock(docker): result 0
```

software-opal replied:
Took me a while to figure out how to get this output; but I think I've managed to fix all the stub errors in the `output.bats` file.

It might be worthwhile adding the `BUILDKITE_AGENT_STUB_DEBUG=/dev/tty` variables commented out in the Docker Compose file for future ~adventurers~ developers 💖.

toote commented on tests/output.bats:
```diff
@@ -46,7 +46,51 @@ load '../lib/run'
+
+  stub buildkite-agent \
+    "meta-data exists docker-compose-plugin-built-image-tag-myservice : exit 1" \
+    "artifact upload : exit 0"
```
here as well:
```
bats-mock(buildkite-agent): got buildkite-agent artifact upload docker-compose-logs/*.log
bats-mock(buildkite-agent): patterns  [2] = 'artifact' 'upload' 
bats-mock(buildkite-agent): arguments [3] = 'artifact' 'upload' 'docker-compose-logs/\*.log' 
bats-mock(buildkite-agent): unexpected argument 'docker-compose-logs/\*.log' at idx 2
bats-mock(buildkite-agent): result 0
```

toote commented on tests/output.bats:
```diff
@@ -29,8 +29,8 @@ load '../lib/run'
 
```
you may want to fix this in the previous line while we are touching these tests:

```bats-mock(buildkite-agent): got buildkite-agent artifact upload docker-compose-logs/*.log
bats-mock(buildkite-agent): patterns  [2] = 'artifact' 'upload' 
bats-mock(buildkite-agent): arguments [3] = 'artifact' 'upload' 'docker-compose-logs/\*.log' 
bats-mock(buildkite-agent): unexpected argument 'docker-compose-logs/\*.log' at idx 2
bats-mock(buildkite-agent): result 0

```

toote reviewed (changes requested):
Just a few minor comments. There is only one important one (to preserve the exit code), all the rest are mostly aesthetic. Sorry for being so nit-picky but I believe that this is the last one as you are obviously did a wonderful job

software-opal commented on commands/run.sh:
```diff
@@ -241,15 +241,35 @@ fi
   echo "~~~ :docker: Starting dependencies"
   if [[ ${#up_params[@]} -gt 0 ]] ; then
-    run_docker_compose "${up_params[@]}" up -d --scale "${run_service}=0" "${run_service}"
+    if ! run_docker_compose "${up_params[@]}" up -d --scale "${run_service}=0" "${run_service}"; then
+      exitcode=1
+    fi
   else
-    run_docker_compose up -d --scale "${run_service}=0" "${run_service}"
+    if ! run_docker_compose up -d --scale "${run_service}=0" "${run_service}"; then
+      exitcode=1
+    fi
   fi
```
I don't believe that the proposed code will preserve the exit code(I think `$?` will be `0`) but I'll check that and put up a change to preserve the exit code shortly.

toote replied:
I may need to stop doing code review past midnight then :P

If you are correct it may be a better idea to do something like `run_docker_compose ... || exitcode=$?`

software-opal replied:
Just checked and `$?` is `0` when you use it inside `if ! false; then`. Using `|| exitcode=$?` is much clearer than the method I had in mind so I've switched to using it.

software-opal commented on commands/run.sh:
```diff
@@ -241,15 +241,35 @@ fi
+
+    return $exitcode
+  fi
 fi
```
I've moved the if-block outside of the dependencies if block. 

As for indenting the rest of the 'run' code; I'm not sure if it would necessarily improve clarity. Additionally the current code layout does allow for easier changes to the code flow if there is a need to add some extra handling when the dependencies fail to start.

toote reviewed (changes requested):
A single minor detail that I caused when suggesting the name change that also caused one of the pipeline steps to fail. I tried to change it myself as it was VERY simple but I didn't have access to push to your branch so it falls to you again. Sorry

software-opal commented on commands/run.sh:
```diff
@@ -239,19 +239,35 @@ elif [[ ! -f "$override_file" ]]; then
+    upload_container_logs "$run_service"
+  fi
+
+  return $exitcode
```
🤦‍♀️, that's what I get for pushing the code without running the tests locally and then walking away from my computer.

software-opal commented:
I'm seeing a README linting failure which I don't think is caused by my code:

```
not ok 3 - Readme version numbers out of date. Latest is v3.11.0
```

Would you like me to change the version numbers in the README to `v3.11.0`?

toote commented:
@software-opal actually, I would suggest you update it to 3.11.1 as that will be the release version after this is merged :)

toote reviewed (approved):
:clap:
